src/devices/src/virtio/net/tap.rs

```diff
@@ -100,11 +116,66 @@ impl IfReqBuilder {
 }
 
 impl Tap {
-    /// Create a TUN/TAP device given the interface name.
+    /// * `if_name` - the name of the interface.
+    /// Create a TUN/TAP device given the tap or macvtap interface name.
     /// # Arguments
     ///
     /// * `if_name` - the name of the interface.
     pub fn open_named(if_name: &str) -> Result<Tap> {
+        // Options:
+        //  - /dev/net/<if_name> exists; open it.
+        //  - It's a macvtap device: determine by checking /sys; open the
+        //    corresponding /dev/tapX node.
+        //  - It's a tap device: open /dev/net/tun and allocate via SETIFF.
+        if let Ok(path) = MacVTap::get_device_node(if_name) {
+            Self::macvtap_open_named(if_name, &path)
+        } else {
+            Self::tap_open_named(if_name)
+        }
+    }
+
+    /// Create a TUN/TAP device given the macvtap interface name and device node.
+    /// # Arguments
+    ///
+    /// * `if_name` - the name of the interface.
+    /// * `dev_path` - location of the interface's device node.
+    fn macvtap_open_named(if_name: &str, dev_path: &Path) -> Result<Tap> {
+        // Open the device node
+        let mut opts = OpenOptions::new();
+        let tap_file = opts
+            .read(true)
+            .write(true)
+            .custom_flags(libc::O_NONBLOCK | libc::O_CLOEXEC)
+            .open(dev_path)
+            .map_err(Error::OpenTapDev)?;
+
+        // Must be a char device
+        let md = tap_file.metadata().map_err(Error::StatTapDev)?;
+        if !md.file_type().is_char_device() {
+            return Err(Error::InvalidTapDevType);
+        }
+
+        // The length check is probably unnecessary because we know it's an extant
```

Comment:
ultra-nit: Although an _extant_ word in English dictionaries, "extant" is not a term I'd use in this comment.

```diff
- // The length check is probably unnecessary because we know it's an extant
+ // The length check is probably unnecessary because we know it's an existent/valid
```